cloud_functions/gcs_event_based_ingest/tests/test_gcs_ocn_bq_ingest_it.py

```diff
@@ -0,0 +1,388 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""unit tests for gcs_ocn_bq_ingest"""
+import json
+import logging
+import os
+import sys
+import uuid
+from time import sleep
+from typing import List
+
+import google.cloud.storage as storage
+import pytest
+from google.cloud import bigquery
+from google.cloud.exceptions import NotFound
+
+sys.path.append(os.path.realpath(os.path.dirname(__file__) + "/.."))
+from gcs_ocn_bq_ingest import main
+
+TEST_DIR = os.path.dirname(os.path.abspath(__file__))
+
+
+@pytest.fixture(scope="module")
+def bq() -> bigquery.Client:
+    """BigQuery Client"""
+    return bigquery.Client(location="US")
+
+
+@pytest.fixture(scope="module")
+def gcs() -> storage.Client:
+    """GCS Client"""
+    return storage.Client()
+
+
+@pytest.fixture(scope="module")
+@pytest.mark.usefixtures("gcs")
+def gcs_bucket(request, gcs) -> storage.bucket.Bucket:
+    """GCS bucket for test artifacts"""
+    bucket = gcs.create_bucket(str(uuid.uuid4()))
+
+    def teardown():
+        bucket.delete(force=True)
+
+    request.addfinalizer(teardown)
+
+    return bucket
+
+
+@pytest.mark.usefixtures("gcs_bucket")
+@pytest.fixture
+def mock_env(gcs, monkeypatch):
+    """environment variable mocks"""
+    # Infer project from ADC of gcs client.
+    monkeypatch.setenv("GCP_PROJECT", gcs.project)
+    monkeypatch.setenv("FUNCTION_NAME", "integration-test")
+
+
+@pytest.mark.usefixtures("bq", "mock_env")
+@pytest.fixture
+def dest_dataset(request, bq, mock_env, monkeypatch):
+    random_dataset = f"test_bq_ingest_gcf_{str(uuid.uuid4())[:8].replace('-','_')}"
+    dataset = bigquery.Dataset(f"{os.getenv('GCP_PROJECT')}"
+                               f".{random_dataset}")
+    dataset.location = "US"
+    bq.create_dataset(dataset)
+    monkeypatch.setenv("BQ_LOAD_STATE_TABLE",
+                       f"{dataset.dataset_id}.serverless_bq_loads")
+    print(f"created dataset {dataset.dataset_id}")
+
+    def teardown():
+        bq.delete_dataset(dataset, delete_contents=True, not_found_ok=True)
+
+    request.addfinalizer(teardown)
+    return dataset
+
+
+@pytest.mark.usefixtures("bq", "mock_env", "dest_dataset")
+@pytest.fixture
+def dest_table(request, bq, mock_env, dest_dataset):
+    with open(os.path.join(TEST_DIR, "resources",
+                           "schema.json")) as schema_file:
+        schema = main.dict_to_bq_schema(json.load(schema_file))
+
+    table = bigquery.Table(
+        f"{os.environ.get('GCP_PROJECT')}.{dest_dataset.dataset_id}.cf_test_nation",
+        schema=schema,
+    )
+
+    table = bq.create_table(table)
+
+    def teardown():
+        bq.delete_table(table, not_found_ok=True)
+
+    request.addfinalizer(teardown)
+    return table
+
+
+@pytest.fixture(scope="function")
+@pytest.mark.usefixtures("gcs_bucket", "dest_dataset", "dest_table")
+def gcs_data(request, gcs_bucket, dest_dataset,
```

Comment:
Where is this referenced?

Reply:
It's a pytest fixture (sort of confusing dynamic syntax) used by these tests:

https://github.com/GoogleCloudPlatform/bigquery-utils/pull/182/files/4a345c8d3aa4068c1c9631cb052cd316f5039c85#diff-0e25c24817556a4a7ba570e46eebf9adf31ad1c091130b4045220a5c54c532f3R132

https://github.com/GoogleCloudPlatform/bigquery-utils/pull/182/files/4a345c8d3aa4068c1c9631cb052cd316f5039c85#diff-0e25c24817556a4a7ba570e46eebf9adf31ad1c091130b4045220a5c54c532f3R157